test/reactors/BaseDutchOrderReactor.t.sol

```diff
@@ -0,0 +1,431 @@
+    DutchOrderReactor,
+    DutchOrder,
+    ResolvedOrder,
+    DutchOutput,
+    DutchInput,
+    BaseReactor
+} from "../../src/reactors/DutchOrderReactor.sol";
+import {OrderInfo, InputToken, OutputToken, SignedOrder} from "../../src/base/ReactorStructs.sol";
+import {OrderQuoter} from "../../src/lens/OrderQuoter.sol";
+import {DutchDecayLib} from "../../src/lib/DutchDecayLib.sol";
+import {CurrencyLibrary} from "../../src/lib/CurrencyLibrary.sol";
+import {NATIVE} from "../../src/lib/CurrencyLibrary.sol";
+import {OrderInfoBuilder} from "../util/OrderInfoBuilder.sol";
+import {MockDutchOrderReactor} from "../util/mock/MockDutchOrderReactor.sol";
+import {MockERC20} from "../util/mock/MockERC20.sol";
+import {DutchOrder, DutchOrderLib} from "../../src/lib/DutchOrderLib.sol";
+import {OutputsBuilder} from "../util/OutputsBuilder.sol";
+import {MockFillContract} from "../util/mock/MockFillContract.sol";
+import {MockFillContractWithOutputOverride} from "../util/mock/MockFillContractWithOutputOverride.sol";
+import {PermitSignature} from "../util/PermitSignature.sol";
+import {ReactorEvents} from "../../src/base/ReactorEvents.sol";
+import {BaseReactorTest} from "../base/BaseReactor.t.sol";
+
+struct TestDutchOrderSpec {
+    uint256 currentTime;
+    uint256 startTime;
+    uint256 endTime;
+    uint256 deadline;
+    DutchInput input;
+    DutchOutput[] outputs;
+}
+
+// Base suite of tests for Dutch decay functionality
+// Intended for extension of reactors that take Dutch style orders
+abstract contract BaseDutchOrderReactorTest is PermitSignature, DeployPermit2, BaseReactorTest {
+    using OrderInfoBuilder for OrderInfo;
+    using DutchOrderLib for DutchOrder;
+
+    OrderQuoter quoter;
+
+    constructor() {
+        quoter = new OrderQuoter();
+    }
+
+    /// @dev Create a signed order and return the order and orderHash
+    /// @param request Order to sign
+    function createAndSignDutchOrder(DutchOrder memory request)
+        public
+        virtual
+        returns (SignedOrder memory signedOrder, bytes32 orderHash)
+    {
+        orderHash = request.hash();
+        return (SignedOrder(abi.encode(request), signOrder(swapperPrivateKey, address(permit2), request)), orderHash);
+    }
+
+    function generateOrder(TestDutchOrderSpec memory spec) private returns (SignedOrder memory order) {
+        vm.warp(spec.currentTime);
+        tokenIn.mint(address(swapper), uint256(spec.input.endAmount));
+        tokenIn.forceApprove(swapper, address(permit2), spec.input.endAmount);
+
+        DutchOrder memory request = DutchOrder({
+            info: OrderInfoBuilder.init(address(reactor)).withDeadline(spec.deadline).withSwapper(address(swapper)),
+            decayStartTime: spec.startTime,
+            decayEndTime: spec.endTime,
+            input: spec.input,
+            outputs: spec.outputs
+        });
+        (order,) = createAndSignDutchOrder(request);
+    }
+
+    function test_dutch_resolveOutputNotStarted() public {
```

Comment:
fyi all these tests come from the validation section of `DutchOrderReactorTest`, but they've been cleaned up and refactored into this abstract test file which can be imported by any reactor test to add coverage to any dutch decaying order types